src/ir/term/mod.rs

```diff
@@ -721,6 +723,104 @@ impl Array {
         self.check_idx(idx);
         self.map.get(idx).unwrap_or(&*self.default).clone()
     }
+
+    /// Iter
+    pub fn into_iter(&self) -> std::collections::btree_map::IntoIter<Value, Value> {
+        self.map.clone().into_iter()
+    }
+}
+
+/// Merge two Array Iterators
+pub struct ArrayMerge {
+    left: Peekable<Box<std::collections::btree_map::IntoIter<Value, Value>>>,
+    right: Peekable<Box<std::collections::btree_map::IntoIter<Value, Value>>>,
+    left_dfl: Value,
+    right_dfl: Value,
+}
+
+impl ArrayMerge {
+    /// Create a new [ArrayMerge] from two [Array]s
+    pub fn new(a: Array, b: Array) -> Self {
+        if a.size != b.size {
+            panic!("IR Arrays have different lengths: {}, {}", a.size, b.size);
+        }
+        if a.key_sort != b.key_sort {
+            panic!(
+                "IR Arrays have different key sorts: {}, {}",
+                a.key_sort, b.key_sort
+            );
+        }
+        if a.default.sort() != b.default.sort() {
+            panic!(
+                "IR Arrays default values have different key sorts: {}, {}",
+                a.default.sort(),
+                b.default.sort()
+            );
+        }
+
+        Self {
+            left: Box::new(a.into_iter()).peekable(),
+            right: Box::new(b.into_iter()).peekable(),
+            left_dfl: *a.default,
+            right_dfl: *b.default,
+        }
+    }
+
+    /// Iter
```

Comment:
This is not right.

1. An iterator should be *lazy*. Here, you're running the whole iterator at once, putting the result into a vector, and then turning that vector into an iterator.
2. As above, `into_iter` is not something you implement as a member function. If you need `into_iter`, you implement the `IntoIterator` trait. But *you do not need to do that here*. More below.

The `ArrayMerge` struct should implement `std::iter::Iterator` directly. There's no need to implement `IntoIterator`. The `next_` function you've written should be part of the `impl Iterator` for `ArrayMerge`, not a member function. And, per the above, it should return something like `Option<(&'a Value, &'a Value, &'a Value)>`.

The functionality inside `next_` seems mostly right, it's just the way that you're connecting the machinery together that isn't right. So fixing this should be pretty easy!

When in doubt, read stdlib code.